avocado/plugins/spawners/lxc.py

```diff
@@ -0,0 +1,221 @@
+
+    def initialize(self):
+        section = "spawner.lxc"
+
+        help_msg = "Distro for the LXC container"
+        settings.register_option(
+            section=section, key="distro", help_msg=help_msg, default="fedora"
+        )
+
+        help_msg = "Release of the LXC container (depends on the choice of distro)"
+        settings.register_option(
+            section=section, key="release", help_msg=help_msg, default="32"
+        )
+
+        help_msg = "Architecture of the LXC container"
+        settings.register_option(
+            section=section, key="arch", help_msg=help_msg, default="i386"
+        )
+
+
+class LXCSpawner(Spawner, SpawnerMixin):
+
+    description = "LXC (container) based spawner"
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+
+    @staticmethod
+    def run_container_cmd(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            exitcode = container.attach_wait(
+                lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err
+            )
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    async def run_container_cmd_async(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            pid = container.attach(
+                lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err
+            )
+            loop = asyncio.get_event_loop()
+            _, exitcode = await loop.run_in_executor(
+                None, os.waitpid, pid, os.WUNTRACED
+            )
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    def is_task_alive(runtime_task):
+        if runtime_task.spawner_handle is None:
+            return False
+
+        container = lxc.Container(runtime_task.spawner_handle)
+        if not container.defined:
+            LOG.debug(f"Container {runtime_task.spawner_handle} is not defined")
+            return False
+        if not container.running:
+            LOG.debug(
+                f"Container {runtime_task.spawner_handle} state is "
+                f"{container.state} instead of RUNNING"
+            )
+            return False
+
+        status, _, _ = LXCSpawner.run_container_cmd(
+            container, ["pgrep", "-r", "R,S", "avocado-runner"]
+        )
+        return status == 0
+
+    async def spawn_task(self, runtime_task):
+        self.create_task_output_dir(runtime_task)
+        task = runtime_task.task
+        # using the "python" symlink will result in the container default python version
+        entry_point_args = ["python", "-m", "avocado.core.nrunner", "task-run"]
+        entry_point_args.extend(task.get_command_args())
+
+        distro = self.config.get("spawner.lxc.distro")
+        release = self.config.get("spawner.lxc.release")
+        arch = self.config.get("spawner.lxc.arch")
+        create_hook = self.config.get("spawner.lxc.create_hook")
```

Comment:
I wasn't able to find out where this config variable is defined and what is its purpose. Can you please give me more info, thanks.

Reply:
Nice catch! This is one of the hooks under question where we can specify a command (e.g. `bash my_customization_script.sh` to run once the containers are created. The idea behind this was to be the first step towards allowing the option (not obligatory though) of creating and customizing containers on the run at least once if they are not available prior. As our approach here is not fully established this is still just a minimal provision for customization though. Would you want me to rename it to something like `spawner.lxc.post_create_hook` or `spawner.lxc.customization_hook`? Indeed as it was added later on, the documentation was lacking, how about
```diff
diff --git a/avocado/plugins/spawners/lxc.py b/avocado/plugins/spawners/lxc.py
index da26c3e4..2fceba6c 100644
--- a/avocado/plugins/spawners/lxc.py
+++ b/avocado/plugins/spawners/lxc.py
@@ -59,6 +59,11 @@ class LXCSpawnerInit(Init):
             section=section, key="arch", help_msg=help_msg, default="i386"
         )
 
+        help_msg = "Container creation hook command to customize optional new containers"
+        settings.register_option(
+            section=section, key="create_hook", help_msg=help_msg, default=""
+        )
+
 
 class LXCSpawner(Spawner, SpawnerMixin):
```
?

Reply:
Ok, I went ahead with this suggestion.